peer/upstream/setup.go

```diff
@@ -109,12 +109,16 @@ func (u *Upstream) Connect(address *util.Connection, serverPublicKey []byte) err
 
 	err := u.client.Connect(address, serverPublicKey, mode.ChainName())
 	if nil == err {
+
+		gc := u.shutdown
+		go u.guaranteeConnection(gc)
```

Comment:
If no particular reason , I think you only need one go routine like Run(). Otherwise, anyone who use connect function will add a go routine.  Maybe use reconnect here,  Isn't it the same purpose here? 
 We need to refactor again for the name Connect (Upstream) , requestConnect (Upstream) and reconnect (socket). This is really confused

Reply:
I want to separate it into specific goroutine, it serves for connection checking. What are you concerning?

Reply:
Correct me if i am wrong, "go u.guaranteeConnection(gc)" is in a function Connect, isn't ?  Whoever wants to use connect will also create a u.guaranteeConnection(gc) routine so there may be many go routines.  Previous code , you put the check in Run(). it has only one go routine so I think you may need only one .

Reply:
If you try to connect with the connected address, it will throws error `address is already in use` so the new goroutine will never been created. But i need to stop it before calling it again. I think it's fine.